pkg/store/bucket.go

```diff
@@ -2303,32 +2351,59 @@ func (r *bucketChunkReader) loadChunks(ctx context.Context, offs []uint32, seq i
 		}
 	}()
 
-	r.chunkBytes = append(r.chunkBytes, b)
 	r.stats.chunksFetchCount++
-	r.stats.chunksFetched += len(offs)
+	r.stats.chunksFetched += len(pIdxs)
 	r.stats.chunksFetchDurationSum += time.Since(fetchBegin)
-	r.stats.chunksFetchedSizeSum += int(end - start)
+	r.stats.chunksFetchedSizeSum += int(part.End - part.Start)
 
-	readOffset := 0
-	for idx, o := range offs {
-		chunkRange := chunkRanges[idx]
+	var (
+		buf        = make([]byte, EstimatedMaxChunkSize)
+		readOffset = int(pIdxs[0].off)
+
+		// Save a few allocations.
+		written  int64
+		diff     uint32
+		chunkLen int
+		n        int
+	)
 
-		// The chunks byte ranges are stored contiguously in the data buffer.
-		cb := (*b)[readOffset : readOffset+chunkRange.length]
-		readOffset += chunkRange.length
+	for i, pIdx := range pIdxs {
+		// Fast forward range reader to the next chunk start in case of sparse (for our purposes) byte range.
+		for readOffset < int(pIdx.off) {
+			written, err = io.CopyN(ioutil.Discard, bufReader, int64(pIdx.off)-int64(readOffset))
+			if err != nil {
+				return errors.Wrap(err, "fast forward range reader")
+			}
+			readOffset += int(written)
+		}
+		// Presume chunk length.
+		chunkLen = EstimatedMaxChunkSize
```

Comment:
What if we are wrong and chunk was larger? (for last chunk)

Reply:
I see now it's done below. Can be a comment here that instead of just `// Presume chunk length.` 

-> 

```
// Presume chunk length. We know the exact length for all but the last chunk. If we understimate we will refetch later.
```

Reply:
Thank you for suggestion, updated comment here.